astronomy/date_body.hpp

```diff
@@ -1041,6 +1067,148 @@ constexpr bool IsValidStretchyUTC(DateTime const& utc) {
+
+// Returns the last entry in [begin, begin + size[ whose UT1 is less than or
+// equal to the given |ut1|.  The range [begin, begin + size[ must be sorted by
+// UT1.
+// We have to use |begin| and |size| rather than |begin| and |end| because
+// otherwise the MSVC complains about undefinedness of |end - begin| (even
+// though Intellisense is fine with it).
+constexpr ExperimentalEOPC02Entry const* LookupUT1(
+    quantities::Time const& ut1,
+    ExperimentalEOPC02Entry const* begin,
+    std::ptrdiff_t const size) {
+  return CHECKING(size > 0,
+                  size == 1
+                      ? CHECKING(begin->ut1_mjd <= mjd(ut1), begin)
+                      : (begin + size / 2)->ut1_mjd <= mjd(ut1)
+                            ? LookupUT1(ut1, begin + size / 2, size - size / 2)
+                            : LookupUT1(ut1, begin, size / 2));
+}
+
+constexpr EOPC04Entry const* LookupUT1(quantities::Time const& ut1,
+                                       EOPC04Entry const* begin,
+                                       std::ptrdiff_t const size) {
+  return CHECKING(size > 0,
+                  size == 1
+                      ? CHECKING(begin->ut1() <= ut1, begin)
+                      : (begin + size / 2)->ut1() <= ut1
+                            ? LookupUT1(ut1, begin + size / 2, size - size / 2)
+                            : LookupUT1(ut1, begin, size / 2));
+}
+
+constexpr ExperimentalEOPC02Entry const* LookupInExperimentalEOPC02(
+    quantities::Time const& ut1) {
+  return LookupUT1(ut1, &experimental_eop_c02[0], experimental_eop_c02.size());
+}
+
+constexpr EOPC04Entry const* LookupInEOPC04(
+    quantities::Time const& ut1) {
+  return LookupUT1(ut1, &eop_c04[0], eop_c04.size());
+}
+
+// Linear interpolation on the UT1 range [low->ut1(), (low + 1)->ut1()].
+constexpr Instant InterpolatedEOPC04(EOPC04Entry const* low,
+                                     quantities::Time const& ut1) {
+  // TODO(egg): figure out whether using the divided difference of the
+  // |p->ut1_minus_tai()|s leads to less catastrophic cancellation than using
+  // the divided difference of the |DateTimeAsUTC(p->utc())|s.
+  return FromTAI(ut1 - (low->ut1_minus_tai() +
+                        (ut1 - low->ut1()) * ((low + 1)->ut1_minus_tai() -
+                                              low->ut1_minus_tai()) /
+                            ((low + 1)->ut1() - low->ut1())));
+}
+
+constexpr Instant InterpolatedExperimentalEOPC02(
+    ExperimentalEOPC02Entry const* low,
+    quantities::Time const& ut1) {
+  return FromTAI(ut1 - (low->ut1_minus_tai +
+                        (mjd(ut1) - low->ut1_mjd) *
+                            ((low + 1)->ut1_minus_tai - low->ut1_minus_tai) /
+                            ((low + 1)->ut1_mjd - low->ut1_mjd)));
+}
+
+// Linear interpolation in the segment between the UT1s |low->ut1_mjd| and
+// |eop_c04.front().ut1()|, used to get continuity when switching between the
+// series.
```

Comment:
The API would be easier to understand if `eop_c04.front()` was passed to the method as a parameter, instead of being accessed as a global.